### ISSUE TITLE
API to trigger immediate sync

### ISSUE THREAD

**Issue by chris13524** [metacontroller/metacontroller#43]
For controllers that depend on external state, Metacontroller has `resyncPeriodSeconds` currently.

I'd like to propose that Metacontroller support an API/endpoint that would trigger an immediate synchronization.

**Comment by hypergig** [comment]
you can have your hook create a Job that sits and curls (or otherwise 'checks in' on) your external resource for changes, when the change you are waiting for happens, end, which will change the status of the Job, which will force metacontroller to resync that controller

**Comment by chris13524** [comment]
@hypergig doesn't that only solve the same thing as `resyncPeriodSeconds`?

**Comment by hypergig** [comment]
not quite, any changes to the parent or children of any controller (composite / decorator) trips the watch on that resource which forces a new call to your hook and a sync

**Comment by chris13524** [comment]
Yes, but triggering that "immediate update" by polling the external resource doesn't really solve the issue here which is to avoid polling by having the thing that changes the external source trigger it.

**Comment by hypergig** [comment]
Oh I sorry, you want a push notification directly from an external thing. You can do that a few ways. You can set up a service account and rbac that allows external access to the parent custom resource. Your external thing (assuming it can curl something on the internet) would patch this object with the changes when they happen. Mutations to the parents and children trip the resync, and you would have those changes (if applicable) in your latest sync, to mutate whatever you need to.

Imperative actions all the way down, no polling. Are you looking for something like that?

**Comment by hypergig** [comment]
Alternatively, if you don't want to use the k8s api, you can employ something like https://github.com/adnanh/webhook.

Just have this act as bastion for the k8s api and security it in anyway you like.

**Comment by chris13524** [comment]
Ok great, and along with `kubectl` w/ a service account I can add some random junk to my parent CRD to trigger the resync. That will work and I will proceed with that in my app.

Leaving this issue open as this may be something to be considered as a feature.

**Comment by hypergig** [comment]
To be honest "add random junk" to the instantiated custom resource object is most likely the most valid answer for your question. The k8s api is after all, the only api metacontroller speaks in this case.

Glad to help

**Comment by grzesuav** [comment]
@chris13524 You have in mind sync on resource level or re-sync the all resources managed by metacontroller ?

**Comment by chris13524** [comment]
@grzesuav I think in my case I had a single resource that I was using, so it wouldn't matter to me if my webhook triggered one resource or all

**Comment by grzesuav** [comment]
https://github.com/metacontroller/metacontroller/issues/43 - @timdevlet moving discussion here

**Comment by grzesuav** [comment]
What I want to clarify is what behavior would be expected here :
* trigger sync of all resources
* trigger sync of all resources managed by specific controller // CRD
* trigger sync of specifc CR
